### PR TITLE
feat: add SYSTEM_OWNER role

### PR DESCRIPTION
**Pull request by eliobischof**
# Why

We need to be able to access some Admin API endpoints with a system user. Instead of copying the endpoints and blowing up the system API, we decided to enable RBAC for system users. This also contributes to our migration towards a resource based API.

# What

This PR enables configuring memberships for system users. Snippet from defaults.yaml (SystemAPIUsers are outcommented in the defaults.yaml):

```yaml
SystemAPIUsers:
# Add keys for authentication of the systemAPI here:
# you can specify any name for the user, but they will have to match the `issuer` and `sub` claim in the JWT:
- superuser:
    Path: /path/to/superuser/ey.pem  # you can provide the key either by reference with the path
    Memberships:
      # MemberType System allows the user to access all APIs for all instances or organizations
      - MemberType: System
        Roles:
          - "SYSTEM_OWNER"
          # Actually, we don't recommend adding IAM_OWNER and ORG_OWNER to the System membership, as this basically enables god mode for the system user
          - "IAM_OWNER"
          - "ORG_OWNER"
      # MemberType IAM and Organization let you restrict access to a specific instance or organization by specifying the AggregateID
      - MemberType: IAM
        Roles: "IAM_OWNER"
        AggregateID: "123456789012345678"
      - MemberType: Organization
        Roles: "ORG_OWNER"
        AggregateID: "123456789012345678"
- superuser2:
    # If no memberships are specified, the user has a membership of type System with the role "SYSTEM_OWNER"
    KeyData: <base64 encoded key>     # or you can directly embed it as base64 encoded value
```

Only membership types *System*, *IAM* and *Organization* are supported.
Calls to any APIs are not limited by quotas if they are authenticated with a system user token.
However, calls with a system user token to other APIs than the system API count against quota usages.

# How

For each request, first the token is checked as before. If the token is invalid, it is checked against the system user jwt profile. If this succeeds, the matching memberships from the config are added to the context for permission checks later.

# TODO

- [x] [Tests](https://github.com/zitadel/zitadel/pull/6778)
- [X] [Docs](https://github.com/zitadel/zitadel/pull/6780)
- [X] [Tech Advisory](https://github.com/zitadel/zitadel/pull/6780)
- [x] [Refactor](https://github.com/zitadel/zitadel/pull/6778)
- [ ] Can we exclude system user calls from authenticated requests? 
- [X] [Read MemberType enum as string from config for better UX](https://github.com/zitadel/zitadel/pull/6780)

### Definition of Ready

- [X] I am happy with the code
- [X] Short description of the feature/issue is added in the pr description
- [X] PR is linked to the corresponding user story -> no story
- [X] Acceptance criteria are met -> no AC
- [X] All open todos and follow ups are defined in a new ticket and justified
- [X] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [X] No debug or dead code
- [X] My code has no repetitions
- [X] Critical parts are tested automatically
- [X] Where possible E2E tests are implemented
- [X] Documentation/examples are up-to-date
- [X] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

File: internal/api/authz/token.go [hunk 1]
```diff
@@ -55,19 +71,21 @@ func Start(authZRepo authZRepo, issuer string, keys map[string]*SystemAPIUser) (
...
-		userID, err := v.verifySystemToken(ctx, token)
-		if err != nil {
-			return "", "", "", "", "", err
-		}
-		return userID, "", "", "", "", nil
-	}
+func (v *TokenVerifier) VerifyAccessToken(ctx context.Context, token string) (userID, clientID, agentID, prefLang, resourceOwner string, isSystemUser bool, err error) {
 	userID, agentID, clientID, prefLang, resourceOwner, err = v.authZRepo.VerifyAccessToken(ctx, token, "", GetInstance(ctx).ProjectID())
-	return userID, clientID, agentID, prefLang, resourceOwner, err
+	if err == nil || !zitadel_errors.IsUnauthenticated(err) {
```
Comment (stebenz) [inline internal/api/authz/token.go:80]: Conditional which mean that either you have no error or you have another error than unauthenticated?
Is this condition meant to be like that?

**Comment by eliobischof** [comment]
@stebenz I tried making it easier to review by creating separate PRs for "peripheric" changes.

- Core changes: #6765 (this PR)
- Tests and major refactorings: #6778 
- Nice enum config: #6780 

Let me know if that helps. If you want to review everything together, we can just merge #6778 and #6780

File: internal/api/authz/token.go [hunk 2]
```diff
@@ -43,7 +44,22 @@ type authZRepo interface {
...
-func Start(authZRepo authZRepo, issuer string, keys map[string]*SystemAPIUser) (v *TokenVerifier) {
+func Start(authZRepo authZRepo, issuer string, keys map[string]*SystemAPIUser) (*TokenVerifier, error) {
+	systemUsers := make(map[string]Memberships, len(keys))
+	for userID, key := range keys {
+		if len(key.Memberships) == 0 {
+			systemUsers[userID] = Memberships{{MemberType: MemberTypeSystem, Roles: []string{"SYSTEM_OWNER"}}}
+			continue
+		}
+		for _, membership := range key.Memberships {
+			switch membership.MemberType {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/api/authz/token.go:55]: I just fixed it here https://github.com/zitadel/zitadel/pull/6780/commits/f0fd5499e990ffcb5854205dfbb77b012a420c7f

File: internal/api/authz/token.go (hunk 1 above)
Comment (eliobischof) [reply internal/api/authz/token.go:80]: Yes. Only if verifying the access token returns an unauthenticated error we want to check if system token verification succeeds.

**Comment by eliobischof** [comment]
> @stebenz I tried making it easier to review by creating separate PRs for "peripheric" changes.
> 
> * Core changes: [feat: add SYSTEM_OWNER role #6765](https://github.com/zitadel/zitadel/pull/6765) (this PR)
> * Tests and major refactorings: [test: system user token test #6778](https://github.com/zitadel/zitadel/pull/6778)
> * Nice enum config: [decode member type string to enum #6780](https://github.com/zitadel/zitadel/pull/6780)
> 
> Let me know if that helps. If you want to review everything together, we can just merge #6778 and #6780

I'll just merge everything together

**Comment by eliobischof** [comment]
@hifabienne is it ok for now if system user requests to other APIs than the system API count as quota usage? It's rather an edge case, I guess. The PR makes sure that these requests are not limited at least.

File: docs/docs/support/advisory/a10007.md
Comment (eliobischof) [inline docs/docs/support/advisory/a10007.md:1]: @hifabienne is this how we write technical advisories?

File: docs/docs/support/advisory/a10007.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+## Statement
+
+This change is tracked in the following PR: [feat: add SYSTEM_OWNER role](https://github.com/zitadel/zitadel/pull/6765).
+As soon as the release version is published, we will include the version here.
+
+## Mitigation
+
+If you have a custom role mapping configured, make sure you configure the new role *SYSTEM_OWNER* before migrating to upcoming ZITADEL versions.
+As a reference, these are the default mappings:
```
Comment (hifabienne) [inline docs/docs/support/advisory/a10007.md:27]: Am I right that those who have no custom mapping, do not have to take any action? Maybe we should clearly state that.

File: docs/docs/support/advisory/a10007.md
Comment (hifabienne) [reply docs/docs/support/advisory/a10007.md:1]: Yes I added a short note to it

**Comment by hifabienne** [comment]
> @hifabienne is it ok for now if system user requests to other APIs than the system API count as quota usage? It's rather an edge case, I guess. The PR makes sure that these requests are not limited at least.

Sorry I don't understand what you mean with this.

File: cmd/defaults.yaml [hunk 4]
```diff
@@ -834,6 +850,28 @@ AuditLogRetention: 0s
...
+        - "system.domain.delete"
+        - "system.debug.read"
+        - "system.debug.write"
+        - "system.debug.delete"
+        - "system.feature.write"
+        - "system.limits.write"
+        - "system.limits.delete"
+        - "system.quota.write"
+        - "system.quota.delete"
+        - "system.iam.member.read"
```
Comment (stebenz) [inline cmd/defaults.yaml:869]: Should system.iam.member.read also be part of SYSTEM_OWNER_VIEWER?

File: cmd/ready/config.go [hunk 5]
```diff
@@ -23,7 +23,6 @@ func MustNewConfig(v *viper.Viper) *Config {
 			mapstructure.StringToTimeDurationHookFunc(),
 			mapstructure.StringToTimeHookFunc(time.RFC3339),
 			mapstructure.StringToSliceHookFunc(","),
-			hook.StringToFeatureHookFunc(),
```
Comment (stebenz) [inline cmd/ready/config.go:26]: Did you intentionally delete the featureHook here?

File: internal/api/authz/context.go [hunk 6]
```diff
@@ -54,29 +58,68 @@ type MemberType int32
...
-	if strings.HasPrefix(method, "/zitadel.system.v1.SystemService") {
-		return CtxData{UserID: userID}, nil
+	userID, clientID, agentID, prefLang, resourceOwner, err := t.VerifyAccessToken(ctx, tokenWOBearer)
+	var sysMemberships Memberships
+	if err != nil {
+		var sysTokenErr error
+		sysMemberships, userID, sysTokenErr = t.VerifySystemToken(ctx, tokenWOBearer, orgID)
+		err = errors.Join(err, sysTokenErr)
+		if sysTokenErr == nil && sysMemberships != nil {
+			err = nil
```
Comment (stebenz) [inline internal/api/authz/context.go:111]: You can delete the next if and do a return here directly

File: internal/api/authz/session_token.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	SessionTokenFormat = SessionTokenPrefix + "%s:%s"
+)
+
+func SessionTokenVerifier(algorithm crypto.EncryptionAlgorithm) func(ctx context.Context, sessionToken, sessionID, tokenID string) (err error) {
+	return func(ctx context.Context, sessionToken, sessionID, tokenID string) (err error) {
+		decodedToken, err := base64.RawURLEncoding.DecodeString(sessionToken)
+		if err != nil {
+			return err
+		}
+		_, spanPasswordComparison := tracing.NewNamedSpan(ctx, "crypto.CompareHash")
```
Comment (stebenz) [inline internal/api/authz/session_token.go:24]: context from the span is not used afterwards?

File: internal/api/authz/session_token.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+		var token string
```
Comment (stebenz) [inline internal/api/authz/session_token.go:25]: necessary to newly declare token as variable? I would just remove the err from the function declaration

File: internal/api/authz/system_token.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	cachedKey, ok := s.cachedKeys[userID]
+	if ok {
+		return &jose.JSONWebKey{KeyID: userID, Key: cachedKey}, nil
+	}
+	key, ok := s.keys[userID]
+	if !ok {
+		return nil, zitadel_errors.ThrowNotFound(nil, "AUTHZ-asfd3", "Errors.User.NotFound")
+	}
+	defer s.mutex.Unlock()
+	s.mutex.Lock()
```
Comment (stebenz) [inline internal/api/authz/system_token.go:110]: lock after defer for unlock

File: internal/api/authz/session_token.go (hunk 7 above)
Comment (eliobischof) [reply internal/api/authz/session_token.go:24]: no, there is no context used anymore after

File: docs/docs/support/advisory/a10007.md (hunk 3 above)
Comment (eliobischof) [reply docs/docs/support/advisory/a10007.md:31]: added it in a new header

File: cmd/ready/config.go (hunk 5 above)
Comment (eliobischof) [reply cmd/ready/config.go:26]: fixed

File: cmd/defaults.yaml (hunk 4 above)
Comment (eliobischof) [reply cmd/defaults.yaml:869]: added to SYSTEM_OWNER_VIEWER

File: internal/api/authz/system_token.go (hunk 9 above)
Comment (eliobischof) [reply internal/api/authz/system_token.go:109]: done

**Comment by eliobischof** [comment]
> > @hifabienne is it ok for now if system user requests to other APIs than the system API count as quota usage? It's rather an edge case, I guess. The PR makes sure that these requests are not limited at least.
> 
> Sorry I don't understand what you mean with this.

Maybe it's a picky detail.

I'm talking about counting requests to limiting or billing usage. I think from a business perspective, authenticated requests coming from a system user should not count to an instances usage. 

The following comment was also about this topic, BTW :slightly_smiling_face: : https://github.com/zitadel/zitadel/pull/6748#issuecomment-1776766214

**Comment by hifabienne** [comment]
> > > @hifabienne is it ok for now if system user requests to other APIs than the system API count as quota usage? It's rather an edge case, I guess. The PR makes sure that these requests are not limited at least.
> > 
> > 
> > Sorry I don't understand what you mean with this.
> 
> Maybe it's a picky detail.
> 
> I'm talking about counting requests to limiting or billing usage. I think from a business perspective, authenticated requests coming from a system user should not count to an instances usage.
> 
> The following comment was also about this topic, BTW 🙂 : [#6748 (comment)](https://github.com/zitadel/zitadel/pull/6748#issuecomment-1776766214)

Ah yes, I think they should not be counted.

**Review by stebenz — DISMISSED** [review]

**Review by stebenz — APPROVED** [review]

**Pull request merged**